### PR TITLE
Fewer ilstubs in corelib from reflection apis

### PR DESCRIPTION
**Pull request by davidwrighton**
Rework use of marshalling for RuntimeTypeHandle, RuntimeModule, RuntimeAssembly, and IRuntimeMethodInfo as used by QCalls
- Remove special QCall only used marshallers for RuntimeAssembly, RuntimeModule and IRuntimeMethodInfo
- Following the pattern of ObjectHandleOnStack, implement QCall handle types for RuntimeAssembly/Module/TypeHandle. Use these in all QCalls that once passed the types directly.
- For uses of IRuntimeMethodInfo, follow the existing RuntimeMethodHandleInternal pattern

Also perform some replacement of bool marshalling with use of Interop.BOOL
- This may trigger build breaks on Linux/Mac, should be easy to fix

Fix delivers a relatively small win on startup, and small throughput gains around reflection as IL stubs are no longer necessary for many functions in reflection that once needed them.

Reduces methods jitted on powershell startup from 422 to 399, (About 5%) but performance win is only about 5ms on ~400ms as the methods removed are simple.

### CODE REVIEW

File: src/System.Private.CoreLib/shared/Interop/Windows/Kernel32/Interop.Globalization.cs [hunk 1]
```diff
@@ -101,8 +101,8 @@ internal static unsafe partial class Kernel32
                     int cchStr);
 
 #if !ENABLE_WINRT
-        [DllImport("Kernel32.dll", CharSet = CharSet.Auto)]
-        internal static extern bool GetUserPreferredUILanguages(uint dwFlags, out uint pulNumLanguages, char[]? pwszLanguagesBuffer, ref uint pcchLanguagesBuffer);
+        [DllImport("kernel32.dll", CharSet = CharSet.Auto)]
+        internal static extern bool GetUserPreferredUILanguages(uint dwFlags, out uint pulNumLanguages, char* pwszLanguagesBuffer, ref uint pcchLanguagesBuffer);
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/Interop/Windows/Kernel32/Interop.Globalization.cs]: This looks like an incomplete fix. We are still going to get a IL stub for this one. Maybe split it into a separate change.

File: src/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 2]
```diff
@@ -373,6 +373,7 @@
     <Compile Include="$(BclSourcesRoot)\System\Variant.cs" />
   </ItemGroup>
   <ItemGroup Condition="'$(TargetsUnix)' == 'true'">
+    <Compile Include="$(BclSourcesRoot)\Interop\Unix\Interop.BOOL.cs" />
```
Comment (jkotas) [inline src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Would it work to just reference the file from `shared`  instead of adding a new copy?

**Comment by jkotas** [comment]
LGTM modulo a few nits.

File: src/System.Private.CoreLib/System.Private.CoreLib.csproj (hunk 2 above)
Comment (davidwrighton) [reply src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Yeah, that would work too. Probably a better solution.

File: src/System.Private.CoreLib/shared/Interop/Windows/Kernel32/Interop.Globalization.cs (hunk 1 above)
Comment (davidwrighton) [reply src/System.Private.CoreLib/shared/Interop/Windows/Kernel32/Interop.Globalization.cs]: I'll just fix the rest of the marshalling. The double call is so cheap as to be meaningless as it isn't repeated regularly so far as I can tell.

**Pull request merged**